server/catgenome/src/main/java/com/epam/catgenome/util/MotifSearchIterator.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2021 EPAM Systems
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.epam.catgenome.util;
+
+import com.epam.catgenome.component.MessageHelper;
+import com.epam.catgenome.constant.MessagesConstants;
+import com.epam.catgenome.entity.reference.motif.Motif;
+import com.epam.catgenome.manager.gene.parser.StrandSerializable;
+import lombok.Value;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class MotifSearchIterator implements Iterator<Motif> {
+    private static final byte CAPITAL_A = 'A';
+    private static final byte CAPITAL_C = 'C';
+    private static final byte CAPITAL_G = 'G';
+    private static final byte CAPITAL_T = 'T';
+    private static final byte CAPITAL_N = 'N';
+    private static final byte LOWERCASE_A = 'a';
+    private static final byte LOWERCASE_C = 'c';
+    private static final byte LOWERCASE_G = 'g';
+    private static final byte LOWERCASE_T = 't';
+    private static final byte LOWERCASE_N = 'n';
+
+    private final LinkedList<Match> positiveMatches = new LinkedList<>();
```

Comment:
could have interface `Queue`:
```
Queue<Match> positiveMatches = new LinkedList<>();
```
it is always better to use interface instead of concrete realization